File: jwt_decoder.go
// jwt_decoder.go
package main

import (
	"encoding/base64"
	"encoding/json"
	"flag"
	"fmt"
	"os"
	"strings"
	"time"
	"crypto/hmac"
	"crypto/sha256"
)

type JWT struct {
	Token          string
	Key            string
	Header         map[string]interface{}
	Payload        map[string]interface{}
	Signature      string
	ValidSignature *bool
	IsExpired      *bool
}

func (j *JWT) Decode() error {
	parts := strings.Split(j.Token, ".")
	if len(parts) != 3 {
		return fmt.Errorf("invalid JWT format")
	}

	// Decode header
	headerBytes, err := base64.RawURLEncoding.DecodeString(parts[0])
	if err != nil {
		return fmt.Errorf("header decode: %v", err)
	}
	if err := json.Unmarshal(headerBytes, &j.Header); err != nil {
		return err
	}

	// Decode payload
	payloadBytes, err := base64.RawURLEncoding.DecodeString(parts[1])
	if err != nil {
		return fmt.Errorf("payload decode: %v", err)
	}
	if err := json.Unmarshal(payloadBytes, &j.Payload); err != nil {
		return err
	}

	j.Signature = parts[2]

	// Check expiration
	if exp, ok := j.Payload["exp"]; ok {
		if expFloat, ok := exp.(float64); ok {
			expTime := time.Unix(int64(expFloat), 0)
			now := time.Now()
			expired := now.After(expTime)
			j.IsExpired = &expired
		}
	} else {
		j.IsExpired = nil
	}

	// Verify signature if key provided
	if j.Key != "" {
		valid, err := j.verifySignature()
		if err == nil {
			j.ValidSignature = &valid
		} else {
			j.ValidSignature = nil
		}
	}
	return nil
}

func (j *JWT) verifySignature() (bool, error) {
	parts := strings.Split(j.Token, ".")
	if len(parts) != 3 {
		return false, fmt.Errorf("invalid JWT")
	}
	data := parts[0] + "." + parts[1]
	mac := hmac.New(sha256.New, []byte(j.Key))
	mac.Write([]byte(data))
	expected := base64.RawURLEncoding.EncodeToString(mac.Sum(nil))
	return expected == parts[2], nil
}

func (j *JWT) ToMap() map[string]interface{} {
	m := map[string]interface{}{
		"header":    j.Header,
		"payload":   j.Payload,
		"signature": j.Signature,
		"token":     j.Token,
	}
	if j.ValidSignature != nil {
		m["valid_signature"] = *j.ValidSignature
	}
	if j.IsExpired != nil {
		m["expired"] = *j.IsExpired
	}
	return m
}

func main() {
	var (
		key      string
		filePath string
		jsonOut  bool
		color    bool
	)
	flag.StringVar(&key, "key", "", "Secret key for HMAC")
	flag.StringVar(&filePath, "file", "", "Read JWT from file")
	flag.BoolVar(&jsonOut, "json", false, "Output as JSON")
	flag.BoolVar(&color, "color", false, "Force color output")
	flag.Parse()

	token := flag.Arg(0)
	if filePath != "" {
		data, err := os.ReadFile(filePath)
		if err != nil {
			fmt.Fprintf(os.Stderr, "Error reading file: %v\n", err)
			os.Exit(1)
		}
		token = strings.TrimSpace(string(data))
	}

	if token == "" {
		// Try stdin
		stat, _ := os.Stdin.Stat()
		if (stat.Mode() & os.ModeCharDevice) == 0 {
			data, _ := os.ReadAll(os.Stdin)
			token = strings.TrimSpace(string(data))
		}
	}
	if token == "" {
		fmt.Fprintln(os.Stderr, "Error: no token provided")
		os.Exit(1)
	}

	jwt := JWT{Token: token, Key: key}
	if err := jwt.Decode(); err != nil {
		fmt.Fprintf(os.Stderr, "Error decoding JWT: %v\n", err)
		os.Exit(1)
	}

	if jsonOut {
		b, _ := json.MarshalIndent(jwt.ToMap(), "", "  ")
		fmt.Println(string(b))
	} else {
		// Print pretty
		fmt.Println("Header:")
		for k, v := range jwt.Header {
			fmt.Printf("  %s: %v\n", k, v)
		}
		fmt.Println("Payload:")
		for k, v := range jwt.Payload {
			fmt.Printf("  %s: %v\n", k, v)
		}
		if jwt.IsExpired != nil {
			if *jwt.IsExpired {
				fmt.Println("  ✗ Token expired")
			} else {
				fmt.Println("  ✓ Token valid")
			}
		}
		if jwt.ValidSignature != nil {
			if *jwt.ValidSignature {
				fmt.Println("  ✓ Signature verified")
			} else {
				fmt.Println("  ✗ Signature invalid")
			}
		}
		fmt.Printf("Signature: %s\n", jwt.Signature)
	}
}

File: jwt_decoder.rs
// jwt_decoder.rs
use serde_json::{json, Value};
use std::fs;
use std::io::{self, Read};
use std::path::Path;
use clap::{App, Arg};
use base64::decode_config;
use base64::URL_SAFE_NO_PAD;
use hmac::{Hmac, Mac};
use sha2::Sha256;
use chrono::{DateTime, Utc};

type HmacSha256 = Hmac<Sha256>;

struct JWT {
    token: String,
    key: Option<String>,
    header: Value,
    payload: Value,
    signature: String,
    valid_signature: Option<bool>,
    expired: Option<bool>,
}

impl JWT {
    fn new(token: String, key: Option<String>) -> Self {
        JWT { token, key, header: Value::Null, payload: Value::Null, signature: String::new(), valid_signature: None, expired: None }
    }

    fn decode(&mut self) -> Result<(), Box<dyn std::error::Error>> {
        let parts: Vec<&str> = self.token.split('.').collect();
        if parts.len() != 3 {
            return Err("Invalid JWT format".into());
        }

        // Decode header
        let header_bytes = decode_config(parts[0], URL_SAFE_NO_PAD)?;
        self.header = serde_json::from_slice(&header_bytes)?;

        // Decode payload
        let payload_bytes = decode_config(parts[1], URL_SAFE_NO_PAD)?;
        self.payload = serde_json::from_slice(&payload_bytes)?;

        self.signature = parts[2].to_string();

        // Check expiration
        if let Some(exp_val) = self.payload.get("exp") {
            if let Some(exp) = exp_val.as_u64() {
                let exp_time = DateTime::<Utc>::from_timestamp(exp as i64, 0).unwrap_or_default();
                self.expired = Some(Utc::now() > exp_time);
            }
        }

        // Verify signature
        if let Some(ref key) = self.key {
            self.valid_signature = Some(self.verify_signature(key)?);
        }

        Ok(())
    }

    fn verify_signature(&self, key: &str) -> Result<bool, Box<dyn std::error::Error>> {
        let parts: Vec<&str> = self.token.split('.').collect();
        let data = format!("{}.{}", parts[0], parts[1]);
        let mut mac = HmacSha256::new_from_slice(key.as_bytes())?;
        mac.update(data.as_bytes());
        let sig_bytes = mac.finalize().into_bytes();
        let sig_b64 = base64::encode_config(sig_bytes, URL_SAFE_NO_PAD);
        Ok(sig_b64 == parts[2])
    }

    fn to_json(&self) -> Value {
        let mut map = json!({
            "header": self.header,
            "payload": self.payload,
            "signature": self.signature,
            "token": self.token,
        });
        if let Some(v) = self.valid_signature { map["valid_signature"] = json!(v); }
        if let Some(v) = self.expired { map["expired"] = json!(v); }
        map
    }

    fn print(&self, color: bool) {
        if color {
            println!("\x1b[32mHeader:\x1b[0m");
            for (k, v) in self.header.as_object().unwrap() {
                println!("  {}: {}", k, v);
            }
            println!("\x1b[33mPayload:\x1b[0m");
            for (k, v) in self.payload.as_object().unwrap() {
                println!("  {}: {}", k, v);
            }
            if let Some(exp) = self.expired {
                if exp {
                    println!("\x1b[31m  ✗ Token expired\x1b[0m");
                } else {
                    println!("\x1b[32m  ✓ Token valid\x1b[0m");
                }
            }
            if let Some(valid) = self.valid_signature {
                if valid {
                    println!("\x1b[32m  ✓ Signature verified\x1b[0m");
                } else {
                    println!("\x1b[31m  ✗ Signature invalid\x1b[0m");
                }
            }
            println!("\x1b[36mSignature: {}\x1b[0m", self.signature);
        } else {
            println!("Header:");
            for (k, v) in self.header.as_object().unwrap() {
                println!("  {}: {}", k, v);
            }
            println!("Payload:");
            for (k, v) in self.payload.as_object().unwrap() {
                println!("  {}: {}", k, v);
            }
            if let Some(exp) = self.expired {
                println!("  Token {}", if exp { "expired" } else { "valid" });
            }
            if let Some(valid) = self.valid_signature {
                println!("  Signature {}", if valid { "verified" } else { "invalid" });
            }
            println!("Signature: {}", self.signature);
        }
    }
}

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let matches = App::new("JWT Decoder")
        .arg(Arg::with_name("token").index(1).help("JWT token"))
        .arg(Arg::with_name("key").long("key").takes_value(true).help("Secret key for HMAC"))
        .arg(Arg::with_name("file").long("file").takes_value(true).help("Read JWT from file"))
        .arg(Arg::with_name("json").long("json").help("Output as JSON"))
        .arg(Arg::with_name("color").long("color").help("Force color output"))
        .get_matches();

    let mut token = matches.value_of("token").map(String::from);
    if let Some(file) = matches.value_of("file") {
        token = Some(fs::read_to_string(file)?.trim().to_string());
    }
    if token.is_none() && !atty::is(atty::Stream::Stdin) {
        let mut s = String::new();
        io::stdin().read_to_string(&mut s)?;
        token = Some(s.trim().to_string());
    }
    if token.is_none() {
        eprintln!("Error: no token provided");
        std::process::exit(1);
    }

    let key = matches.value_of("key").map(String::from);
    let mut jwt = JWT::new(token.unwrap(), key);
    if let Err(e) = jwt.decode() {
        eprintln!("Error decoding JWT: {}", e);
        std::process::exit(1);
    }

    if matches.is_present("json") {
        println!("{}", serde_json::to_string_pretty(&jwt.to_json())?);
    } else {
        let color = matches.is_present("color") || atty::is(atty::Stream::Stdout);
        jwt.print(color);
    }
    Ok(())
}
